ivy/__init__.py

```diff
@@ -1378,3 +1143,173 @@ def cast_data_types(val=True):
     upcast_dtypes = val
     downcast_dtypes = val
     crosscast_dtypes = val
+
+
+# Promotion Tables #
+# ---------------- #
+
+
+# data type promotion
+array_api_promotion_table = {
+    (uint16, int16): int32,
+    (uint32, int8): int64,
+    (bool, bool): bool,
+    (uint16, uint64): uint64,
+    (uint8, int16): int16,
+    (uint32, int32): int64,
+    (uint16, int64): int64,
+    (uint64, uint64): uint64,
+    (float32, float32): float32,
+    (int8, int16): int16,
+    (uint16, uint16): uint16,
+    (uint8, uint64): uint64,
+    (float16, float64): float64,
+    (int32, int64): int64,
+    (uint8, int64): int64,
+    (uint16, int8): int32,
+    (float16, float16): float16,
+    (uint8, uint16): uint16,
+    (int16, int16): int16,
+    (int8, int64): int64,
+    (uint16, int32): int32,
+    (uint8, int8): int16,
+    (uint32, uint32): uint32,
+    (int32, int32): int32,
+    (int8, int8): int8,
+    (float32, float64): float64,
+    (uint8, int32): int32,
+    (int16, int64): int64,
+    (int8, int32): int32,
+    (int64, int64): int64,
+    (uint8, uint8): uint8,
+    (uint32, int16): int64,
+    (uint16, uint32): uint32,
+    (int16, int32): int32,
+    (uint32, uint64): uint64,
+    (uint32, int64): int64,
+    (uint8, uint32): uint32,
+    (float64, float64): float64,
+    (float16, float32): float32,
+}
+
+# the extra promotion table follows numpy safe casting convention
+# the following link discusses the different approaches to dtype promotions
+# https://jax.readthedocs.io/en/latest/jep/9407-type-promotion.html
+common_extra_promotion_table = {
+    (float32, complex128): complex128,
```

Comment:
Just to be sure, we're also explicitly maintaining the reverse op too right? Maybe let's remove the duplicates in the promotion table in a separate PR after discussing with @djl11